service/frontend/nexus_incoming_service_registry.go

```diff
@@ -0,0 +1,406 @@
+	listNexusIncomingServicesLongPollMinWait dynamicconfig.DurationPropertyFn,
+	listNexusIncomingServicesRetryPolicy backoff.RetryPolicy,
+	matchingClient matchingservice.MatchingServiceClient,
+	persistence p.NexusIncomingServiceManager,
+) NexusIncomingServiceRegistry {
+	return &registry{
+		status:                         common.DaemonStatusInitialized,
+		serviceDataReady:               future.NewFuture[struct{}](),
+		refreshServicesLongPollTimeout: listNexusIncomingServicesLongPollTimeout,
+		refreshServicesLongPollMinWait: listNexusIncomingServicesLongPollMinWait,
+		refreshServicesRetryPolicy:     listNexusIncomingServicesRetryPolicy,
+		matchingClient:                 matchingClient,
+		persistence:                    persistence,
+	}
+}
+
+func (r *registry) Start() {
+	if !atomic.CompareAndSwapInt32(&r.status, common.DaemonStatusInitialized, common.DaemonStatusStarted) {
+		return
+	}
+
+	refreshCtx := headers.SetCallerInfo(
+		context.Background(),
+		headers.SystemBackgroundCallerInfo,
+	)
+
+	r.refreshPoller = goro.NewHandle(refreshCtx).Go(r.refreshServices)
+}
+
+func (r *registry) Stop() {
+	if !atomic.CompareAndSwapInt32(&r.status, common.DaemonStatusStarted, common.DaemonStatusStopped) {
+		return
+	}
+
+	r.refreshPoller.Cancel()
+	<-r.refreshPoller.Done()
+}
+
+func (r *registry) CreateOrUpdateNexusIncomingService(
+	ctx context.Context,
+	request *operatorservice.CreateOrUpdateNexusIncomingServiceRequest,
+) (*operatorservice.CreateOrUpdateNexusIncomingServiceResponse, error) {
+	resp, err := r.matchingClient.CreateOrUpdateNexusIncomingService(ctx, &matchingservice.CreateOrUpdateNexusIncomingServiceRequest{
+		Service: request.Service,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return &operatorservice.CreateOrUpdateNexusIncomingServiceResponse{Service: entryToService(resp.Entry)}, nil
+}
+
+func (r *registry) DeleteNexusIncomingService(
+	ctx context.Context,
+	request *operatorservice.DeleteNexusIncomingServiceRequest,
+) (*operatorservice.DeleteNexusIncomingServiceResponse, error) {
+	_, err := r.matchingClient.DeleteNexusIncomingService(ctx, &matchingservice.DeleteNexusIncomingServiceRequest{
+		Name: request.Name,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return &operatorservice.DeleteNexusIncomingServiceResponse{}, nil
+}
+
+func (r *registry) GetNexusIncomingService(
+	ctx context.Context,
+	request *operatorservice.GetNexusIncomingServiceRequest,
+) (*operatorservice.GetNexusIncomingServiceResponse, error) {
+	if !r.isTableVersionCurrent(ctx) {
+		r.invalidateCachedServices()
+		if err := r.fetchServices(ctx); err != nil {
+			return nil, fmt.Errorf("error loading nexus incoming services frontend cache: %w", err)
+		}
+	}
+
+	r.RLock()
+	defer r.RUnlock()
+
+	service, ok := r.servicesByName[request.Name]
+	if !ok {
+		return nil, serviceerror.NewNotFound(fmt.Sprintf("frontend could not find cached service with name %v", request.Name))
```

Comment:
This is a user facing error, don't give away too many details.
"frontend" and "cached" will confuse the average end user.